pkg/apis/pipeline/v1alpha1/gcs_resource.go

```diff
@@ -0,0 +1,172 @@
+	if r.Spec.Type != PipelineResourceTypeStorage {
+		return nil, fmt.Errorf("GCSResource: Cannot create a GCS resource from a %s Pipeline Resource", r.Spec.Type)
+	}
+	var location string
+	var locationSpecified, dir bool
+
+	for _, param := range r.Spec.Params {
+		switch {
+		case strings.EqualFold(param.Name, "Location"):
+			location = param.Value
+			if param.Value != "" {
+				locationSpecified = true
+			}
+		case strings.EqualFold(param.Name, "Dir"):
+			dir = true // if dir flag is present then its a dir
+		}
+	}
+
+	if !locationSpecified {
+		return nil, fmt.Errorf("GCSResource: Need Location to be specified in order to create GCS resource %s", r.Name)
+	}
+	return &GCSResource{
+		Name:     r.Name,
+		Type:     r.Spec.Type,
+		Location: location,
+		TypeDir:  dir,
+		Secrets:  r.Spec.SecretParams,
+	}, nil
+}
+
+// GetName returns the name of the resource
+func (s GCSResource) GetName() string {
+	return s.Name
+}
+
+// GetType returns the type of the resource, in this case "storage"
+func (s GCSResource) GetType() PipelineResourceType {
+	return PipelineResourceTypeStorage
+}
+
+// GetParams get params
+func (s *GCSResource) GetParams() []Param { return []Param{} }
+
+// GetSecretParams returns the resource secret params
+func (s *GCSResource) GetSecretParams() []SecretParam { return s.Secrets }
+
+// Replacements is used for template replacement on an GCSResource inside of a Taskrun.
+func (s *GCSResource) Replacements() map[string]string {
+	return map[string]string{
+		"name":     s.Name,
+		"type":     string(s.Type),
+		"location": s.Location,
+	}
+}
+
+// SetDestinationDirectory sets the destination directory at runtime like where is the resource going to be copied to
+func (s *GCSResource) SetDestinationDirectory(destDir string) { s.DestinationDir = destDir }
+
+// GetUploadContainerSpec gets container spec for gcs resource to be uploaded like
+// set environment variable from secret params and set volume mounts for those secrets
+func (s *GCSResource) GetUploadContainerSpec() ([]corev1.Container, error) {
+	if s.DestinationDir == "" {
+		return nil, fmt.Errorf("GCSResource: Expect Destination Directory param to be set: %s", s.Name)
+	}
+	var args []string
+	if s.TypeDir {
+		args = []string{"-args", fmt.Sprintf("cp -r %s %s", filepath.Join(s.DestinationDir, "*"), s.Location)}
+	} else {
+		args = []string{"-args", fmt.Sprintf("cp %s %s", filepath.Join(s.DestinationDir, "*"), s.Location)}
+	}
+
+	envVars, secretVolumeMount := getSecretEnvVarsAndVolumeMounts(s.Name, s.Secrets)
+
+	return []corev1.Container{{
+		Name:         fmt.Sprintf("storage-upload-%s", s.Name),
+		Image:        *gsutilImage,
+		Args:         args,
+		VolumeMounts: secretVolumeMount,
+		Env:          envVars,
+	}}, nil
+}
+
+// GetDownloadContainerSpec returns an array of container specs to download gcs storage object
+func (s *GCSResource) GetDownloadContainerSpec() ([]corev1.Container, error) {
```

Comment:
oooo i like this UploadContainer + DownloadContainer pattern! do you see us doing this for other resource types as well? (e.g. git)

Reply:
Yes I did not want to do git container spec in this PR but it fits the case well. s3 resource would be a great fit too.

Reply:
kk sounds great!